### ISSUE TITLE
test-unit failing tests

### ISSUE THREAD

**Issue by gogainda** [oracle/truffleruby#2059]
```
========================================================================================================================
<internal:core> core/throw_catch.rb:36:in `catch'
<internal:core> core/throw_catch.rb:36:in `catch'
      94:             :location  => normalize_location(fault.location),
      95:           }
      96:         end
  =>  97:         assert_equal([
      98:                        {
      99:                          :class     => Failure,
     100:                          :message   => "failure",
/Users/novoi/tmp/test-unit-master/test/test-test-case.rb:97:in `test_add_failed_assertion'
<[{:class=>Test::Unit::Failure,
  :location=>
   ["/Users/novoi/tmp/test-unit-master/test/test-test-case.rb:0:in `test_failure'",
    "/Users/novoi/tmp/test-unit-master/test/test-test-case.rb:0:in `test_add_failed_assertion'"],
  :message=>"failure",
  :test_name=>"test_failure(TC_FailureError)"}]> expected but was
<[{:class=>Test::Unit::Failure,
  :location=>
   ["/Users/novoi/tmp/test-unit-master/test/test-test-case.rb:0:in `test_failure'",
    "<internal:core> core/throw_catch.rb:0:in `catch'",
    "/Users/novoi/tmp/test-unit-master/test/test-test-case.rb:0:in `test_add_failed_assertion'",
    "<internal:core> core/throw_catch.rb:0:in `catch'",
    "<internal:core> core/throw_catch.rb:0:in `catch'"],
  :message=>"failure",
  :test_name=>"test_failure(TC_FailureError)"}]>

diff:
  [{:class=>Test::Unit::Failure,
    :location=>
     ["/Users/novoi/tmp/test-unit-master/test/test-test-case.rb:0:in `test_failure'",
+     "<internal:core> core/throw_catch.rb:0:in `catch'",
?     "/Users/novoi/tmp/test-unit-master/test/test-test-case.rb:0:in `test_add_failed_assertion'"],
+     "<internal:core> core/throw_catch.rb:0:in `catch'",
+     "<internal:core> core/throw_catch.rb:0:in `catch'"],
    :message=>"failure",
    :test_name=>"test_failure(TC_FailureError)"}]
Failure: test_add_failed_assertion(Test::Unit::TestTestCase)
```
To reproduce:
```
git clone git@github.com:test-unit/test-unit.git
bundle install
rake
```

**Comment by eregon** [comment]
I think in this case it's probably up to test-unit to filter out `<internal:` backtrace entries, but it would be good to investigate a bit to be sure, before.
The `core/throw_catch.rb:0` looks a bit weird (there is no line 0, lines start at 1).

**Comment by eregon** [comment]
I checked and I believe it makes sense to fix this in test-unit:
https://github.com/test-unit/test-unit/blob/ea81cf9545368ca9fae704454fe67d81ad9fe2c9/test/test-test-case.rb#L65-L68 needs to also ignore `<internal:` backtrace entries.
I filed https://github.com/test-unit/test-unit/issues/171.

**Comment by eregon** [comment]
Let's reopen this to track the two remaining failing tests when running test-unit's tests:
https://travis-ci.org/github/test-unit/test-unit/jobs/714613316
```
 $bundle exec rake
...
............................................................................F
============================================================================================================================================
<internal:core> core/throw_catch.rb:36:in `catch'
<internal:core> core/throw_catch.rb:36:in `catch'
     850:                 :line => line_after_def,
     851:                 :method_name => "test_before",
     852:               }
  => 853:               assert_false(test_case.test_defined?(query))
     854:             end
     855:           end
     856: 
/home/eregon/code/test-unit/test/test-test-case.rb:853:in `test_line_after_def'
Failure: test_line_after_def(Test::Unit::TestTestCase::TestTestDefined::TestCombine::TestDefStyle):
  <false> expected but was
  <true>
============================================================================================================================================
....F
============================================================================================================================================
<internal:core> core/throw_catch.rb:36:in `catch'
<internal:core> core/throw_catch.rb:36:in `catch'
     661:                 def test_nothing
     662:                 end
     663:               end
  => 664:               assert_false(test_case.test_defined?(:line => line_before))
     665:             end
     666: 
     667:             def test_def
/home/eregon/code/test-unit/test/test-test-case.rb:664:in `test_before'
Failure: test_before(Test::Unit::TestTestCase::TestTestDefined::TestLine::TestDefStyle):
  <false> expected but was
  <true>
============================================================================================================================================
....................................O
...
```

**Comment by eregon** [comment]
@bjfish Can you take look at those?

**Comment by bjfish** [comment]
@eregon I think this small reproduction case recreates the issue:
```
class TestCase
    def self.method_added(name)
        location = caller_locations(1, 1)[0]
        puts "location.lineno #{location.lineno}"
    end
end

Class.new(TestCase) do
    def test_method # line 9
    end
end

```
MRI 2.6.6
```
location.lineno 9
```
TruffleRuby
```
location.lineno 8
```

**Comment by eregon** [comment]
I see, I think we've seen this before. We probably need to pass the `AddMethodNode` as the "call site" for calling `method_added` inside `ModuleFields#addMethod`.

I think one way to do is to set `EncapsulatingNodeReference` around calling `method_added`.
We could also pass a DirectCallNode or IndirectCallNode to `ModuleFields#addMethod` but that seems heavy for something that's normally just called once.

**Comment by eregon** [comment]
I've got a fix for this but it needs some extra fixes in Truffle.

**Comment by eregon** [comment]
Fixed by 953ac5d2c62586de8f4b0a2cf9e0fad9de78cea0, all test-unit tests pass locally for me now.
